minject/registry.py

```diff
@@ -238,28 +309,93 @@ def get(
+        return _unwrap(self._get_by_metadata(meta, default))
 
-            obj_list = self._by_iface.get(key)
-            if obj_list:
-                return _unwrap(obj_list[0])
+    async def aget(self, key: "RegistryKey[T]") -> Optional[T]:
+        """
+        Resolve objects marked with the @async_context decorator.
+        """
+        if not _is_key_async(key):
+            raise AssertionError("key must be async to use aget")
 
-        return _unwrap(self._get_by_metadata(meta, default))
+        if not self._async_entered:
+            raise AssertionError("cannot use aget outside of async context")
+
+        meta = _get_meta_from_key(key)
+        maybe_class = self._get_if_already_in_registry(key, meta)
+        if maybe_class is not None:
+            return maybe_class
+
+        by_meta = await self._aget_by_metadata(meta)
+        return _unwrap(by_meta)
+
+    def _get_if_already_in_registry(
+        self, key: "RegistryKey[T]", meta: "RegistryMetadata[T]"
+    ) -> Optional[T]:
+        # retrieve the class metdata, and the metadata of class
+        # without inherited metadata
+        meta = _get_meta_from_key(key)
+
+        # if the class has already been registered, return it
+        if meta in self._by_meta:
+            return _unwrap(self._by_meta[meta])
+
+        # following checks only apply if key is a class
+        if not isinstance(key, type):
+            return None
+
+        # If the class has no metadata, but a parent has been registered in registry,
+        # return the registered parent. If the class has metadata, we should not
+        # check parents, we must use the metadata attached to the class to
+        # construct the object.
+        meta_no_bases = _get_meta(key, include_bases=False)
+        obj_list = self._by_iface.get(key)
+        if meta_no_bases is None and obj_list:
+            return _unwrap(obj_list[0])
+
+        # nothing has been registered for this metadata yet
+        return None
+
+    async def __aenter__(self) -> "Registry":
+        """
+        Mark a registry instance as ready for resolving async objects.
+        """
+        if self._async_entered:
+            raise AssertionError(
+                "Attempting to enter registry context while already in context. This should not happen."
+            )
+        self._async_entered = True
+        return self
+
+    async def __aexit__(self, exc_type, exc_value, traceback) -> None:
+        """
+        Closes the registry. Closes all contexts on the registry's context stack
+        and then closes the registry itself with regisry.close().
+        """
+        if not self._async_entered:
+            raise AssertionError(
+                "Attempting to exit registry context while not in context. This should not happen."
+            )
+        self._async_entered = False
+        # close all objects in the registry
+        await self._async_context_stack.aclose()
```

Comment:
What happens if any of the `__aexit__` methods on the stack ends up raising an exception? (Does the call to `self.close` need to go in a `finally:` block?)

Reply:
Great catch, agree this should be in a finally block. I'll add a test for this now.